import-scripts/precache-portals.py

```diff
@@ -0,0 +1,147 @@
+from dateutil.parser import parse
+from multiprocessing.dummy import Pool
+import argparse
+import datetime
+import json
+import os
+import requests
+import subprocess
+import sys
+import time
+
+DEPLOYMENTS = ["cbioportal-backend-genie-archive",
+                "cbioportal-backend-genie-public",
+                "cbioportal-backend-genie-private",
+                "cbioportal-backend-master"]
+
+DEPLOYMENT_TO_URL_MAP = {"cbioportal-backend-genie-archive":"genie-archive", 
+                         "cbioportal-backend-genie-public":"genie",
+                         "cbioportal-backend-genie-private":"genie-private"
+                        }
+
+PROTECTED_PORTALS=["cbioportal-backend-genie-public",
+                    "cbioportal-backend-genie-private",
+                    "cbioportal-backend-genie-private"]
+
+MINUTES_BETWEEN_CHECK = 5
+MAX_NUMBER_OF_ATTEMPTS = 18
+# ----------------------------------------------------------
+# Functions wrapping k8s/timing checks
+
+# get pods associated with deployment
+def get_deployed_pods(deployment_name):
+    deployed_pods = subprocess.check_output(["/bin/bash", "-c", "kubectl get pods | grep %s | sed 's/\s\s*/\t/g' | cut -f1" % (deployment_name)]).rstrip().split("\n")
+    return deployed_pods
+
+# return datetime object representing last deployment restart
+def get_last_deployment_restart_datetime(deployment_name):
+    last_restart_time =  subprocess.check_output(["/bin/bash", "-c", "kubectl describe deployment %s | grep 'LAST_RESTART'" % (deployment_name)]).strip().split()[2:]
+    return parse(" ".join(last_restart_time))
+
+def get_last_pod_start_datetime(deployment):
+    pod_names = get_deployed_pods(deployment)
+    pod_start_datetimes = []
+    for pod_name in pod_names:
+        pod_start_time = subprocess.check_output(["/bin/bash", "-c", "kubectl describe pod %s | grep -C1 'State.*Running' | grep 'Started'" % (pod_name)]).strip().split()[1:]
+        pod_start_datetimes.append(parse(" ".join(pod_start_time)))
+    return sorted(pod_start_datetimes, reverse = True)[0]
+
+# check if restart occurred since last restart (minute offset should be set according to how frequently script is ran)
+# use date command so timezone is auto-included
+def restart_occurred_since_last_check(last_deployment_restart_datetime):
+    current_datetime = parse(subprocess.check_output(["/bin/bash", "-c", "date +'%b %d %H:%M:%S %z %Y'"]))
+    last_check_datetime = current_datetime - datetime.timedelta(minutes=MINUTES_BETWEEN_CHECK)
```

Comment:
time of last check is calculated here based on crontab cycle .. if we adjust crontab we must remember to adjust this constant.  Potentially we could get the MINUTE_BETWEEN_CHECK as an argument --- to help us remember to update it when we update crontab. Alternatively, we could store the time of last check in a file (rather than computing it). But we could go ahead with this as is.